docs(state) add select and stateful tests, add docs, update readme (#82)
BioPhoton/rx-angular: libs/state/docs/snippets/hostbindings.md

# HostBindings

Some examples how to reactively handle basic [`HostBindings`](https://angular.io/api/core/HostBinding) with `@rx-angular/state` `RxState`.

Sadly `HostBindings` are not able to bind to `Observable` sources out of the box. So we have to come up with custom solutions

> in order still have fully reactive components.

Author: ```suggestion
in order to have fully reactive components.
```

In the following examples we will use the `rxLet` directive or the `push` pipe as replacements for angulars `async` pipe.
`rxLet` and `push` belong to the not yet released `@rx-angular/template` package.

Furthermore we want to express that we will come up with a more convenient solution facing this problem. This can be seen as WIP and
should not be the longterm solution to handle `HostBindings` in a fully reactive way.

Imagine you have the following state which you want to bind to properties of your host element.

```ts
interface ComponentState {
  visible: boolean;
  top: number;
  maxHeight: number;
}
```

## Be aware of changeDetection - HostBindings are not reactive

In this setup we assign our `HostBindings` to the `get()` method of our state.

As stated in the title, we have to be aware changeDetection. On every changeDetection cycle, angular will re-evaluate
all `HostBindings`. If our component doesn't get flagged as dirty, our `HostBindings` won't get updated. So we have to make

> sure that our state changes which should update our `HostBindings` are actually resulting in changeDetection.

Author: ```suggestion
sure that state changes that are related to the `HostBindings` value are actually triggering a re-render.
```

```ts
@Component({
  providers: [RxState]
})
export class RxComponent {

>   @HostBinding('[class.is-hidden]') get isHidden() {

Author: ```suggestion
  // Modifiing the calss
  @HostBinding('[class.is-hidden]') get isHidden() {
```

    return !this.state.get().visible;
  }

>   @HostBinding('[style.marginTop]') get marginTop() {

Author: ```suggestion
  // Modifying styles
  @HostBinding('[style.marginTop]') get marginTop() {
```

>     return `${this.state.get().top ? this.state.get().top : 0}px`;

Author: ```suggestion
    return `${this.state.get().top}px`;
```

  }

>   @HostBinding('[style.maxHeight]') get maxHeight() {

Author: ```suggestion
  // Modifiing styles
  @HostBinding('[style.maxHeight]') get maxHeight() {
```

>     return `${this.state.get().maxHeight ? this.state.get().maxHeight : 0}px`;

Author: ```suggestion
    return `${this.state.get().maxHeight}px`;
```

  }

  constructor(private state: RxState<ComponentState>) {}
}
```

With this setup in place we have two options to get things done.

### Call ChangeDetection manually

> Since rendering is a side-effect in state management terms, we could utilize the `hold` method and register

Author: ```suggestion
Since rendering is a side-effect, we could utilize the `hold` method and register
```

> a side-effect which handles changeDetection for us.

Author: ```suggestion
a function which handles change detection for us.
```

```ts
@Component({
  providers: [RxState]
})
export class RxComponent {

>   @HostBinding('[class.is-hidden]') get isHidden() {

Author: ```suggestion
  // Modifying the class
  @HostBinding('[class.is-hidden]') get isHidden() {
```

    return !this.state.get().visible;
  }

>   @HostBinding('[style.marginTop]') get marginTop() {

Author: ```suggestion
  @HostBinding('[style.marginTop]') get marginTop() {
```
```suggestion
  // Modifying styles
  @HostBinding('[style.marginTop]') get marginTop() {
```

>     return `${this.state.get().top ? this.state.get().top : 0}px`;

Author: ```suggestion
    return `${this.state.get().top}px`;
```

  }

>   @HostBinding('[style.maxHeight]') get maxHeight() {
>     return `${this.state.get().maxHeight ? this.state.get().maxHeight : 0}px`;

Author: ```suggestion
  // Modifying styles
  @HostBinding('[style.maxHeight]') get maxHeight() {
    return `${this.state.get().maxHeight}px`;
```
